docker/pypi/wmagent/run.sh

```diff
@@ -0,0 +1,683 @@
+        return $(false)
+    fi
+
+    # Update flavor/type/domain global variables if needed
+    # (grep -E "^[[:blank:]]*ORACLE_" $WMA_ADMIN_DIR/WMAgent.secrets > /dev/null) && CURR_FLAVOR=oracle || CURR_FLAVOR=mysql
+
+    echo "Done: $stepMsg"
+    echo "-------------------------------------------------------"
+}
+
+_check_oracle()
+{
+    # Auxiliary finction to check if the oracle database configured for the current agent is empty
+    # NOTE: Oracle is centraly provided - we require an empty database for every account/agent
+    #       otherwise we cannot guarantie this is the only agent to connect to the so configured database
+    echo "$FUNCNAME: Checking whether the oracle database is clean and not used by other agents ..."
+    $manage db-prompt<<"    EOF">/tmp/db_check_output
+    SELECT COUNT(*) from USER_TABLES;
+    EOF
+    tables=`cat /tmp/db_check_output | grep -A1 '\-\-\-\-' | tail -n 1`
+    rm -rf /tmp/db_check_output
+    if [ "$tables" -gt 0 ]; then
+        echo "ERROR: Non empty database found: $tables tables."; return $(false)
+    else
+        echo "OK"; return $(true)
+    fi
+}
+
+_check_mysql()
+{
+    # Auxiliary finction to check if the the MariaDB database for the current agent is properly set
+    # TODO: To be implemented in the issue related to the MariaDB setup fro wmagent
+    echo "$FUNCNAME: Checking whether the mysql schema has been installed"
+    true
+}
+
+check_databases() {
+    # TODO: Here to check all databases - relational and CouchDB
+    #       * call check_oracle or check_sql or similar
+    #       * call check_couchdb
+    local oracleCred=false
+    local mysqlCred=false
+    (grep -E "^[[:blank:]]*(ORACLE_USER)" $WMA_ADMIN_DIR/WMAgent.secrets > /dev/null) && \
+        (grep -E "^[[:blank:]]*(ORACLE_PASS)" $WMA_ADMIN_DIR/WMAgent.secrets > /dev/null) && \
+        (grep -E "^[[:blank:]]*(ORACLE_TNS)" $WMA_ADMIN_DIR/WMAgent.secrets > /dev/null) && \
+        oracleCred=true
+
+    (grep -E "^[[:blank:]]*(MYSQL_USER)" $WMA_ADMIN_DIR/WMAgent.secrets > /dev/null) && \
+        (grep -E "^[[:blank:]]*(MYSQL_PASS)" $WMA_ADMIN_DIR/WMAgent.secrets > /dev/null) && \
+        mysqlCred=true
+
+    # Checking for relational database credentials at WMAgent.secrets file
+    case $FLAVOR in
+        mysql)
+            $mysqlCred   || { echo "ERROR: No Mysql database credentials provided at $WMA_ADMIN_DIR/WMAgent.secrets"; return $(false) ;}
+            _check_mysql || { echo "ERROR: MaridDB database unreachable or not cleaned"; return $(false) ;}
+            ;;
+        oracle)
+            $oracleCred   || { echo "ERROR: No Oracle database credentials provided at $WMA_ADMIN_DIR/WMAgent.secrets"; return $(false) ;}
+            _check_oracle || { echo "ERROR: Oracle database unreachable or not cleaned"; return $(false) ;}
+        ;;
+    esac
+}
+
+check_docker_init() {
+    # A function to check all previously populated */.dockerInit files
```

Comment:
I am afraid I still do not have a thorough understanding of this docker build id vs docker init files.

Let me try with a real scenario: what would happen if we want to run a new wmagent container with one or two configuration lines that have been changed? Would the container start with the new configurations? Would it fail to start? Or what?

Reply:
The general principal is:  
* At build time - we create the unique build id and we record it in the image 
* At run time - on every initialization step we search for the respective `.dockerInit` file in the host mount points and compare its contents against the `buildId` inside the image. If we do not find such file or the so recorded hash inside differs from the `buildId` of the current image, then we consider this agent has either never passed through the relevant initialization step or what we find is a leftover from a previous image. Then we trigger the initialization step and if successful at the end we record the new `.dockerInit` file containing the current `buildId`. This way we guarantee we never repeat an initialization step. 
* In addition, we use `WMAgent.secrets` file as a sole source for generating the actual agent config file. In order to check for any configuration (not initialization) change, during the agent initialization we create the `WMAgent.secret`'s file md5 sum. And on every container/agent restart we compare it. If a change has happened there, we delete only `$WMA_CONFIG_DIR/.dockerInit` and `$WMA_HOSTADMIN_DIR/.dockerInit`, which are the initialization steps that could be affected by a change in this file (one of them checks the validity of the `WMAgent.secrets` file, the other regenerates the agent config). They will be the only initialization steps triggered again then.

Now to answer your question. 
* If the configuration change has happened in `WMAGent.secrets` file, then the above mechanism for re-configuring the agent will be triggered automatically. And the configuration steps will be reinitialized.
* If the change has happened in the agent config directly, we have no mechanism to reflect it in anyway. So if this is a configuration change that affects other parts of the configuration as well (like database reconfiguration etc., which I cannot imagine how could be the case, but anyway), in such cases we may want to repeat the whole configuration step from scratch. Then somebody needs to go there and manually delete `$WMA_CONFIG_DIR/.dockerInit` and restart the agent.

Honestly I do not think we may have a configuration change in the agent config file itself such that would affect other parts beyond the agent's executable itself. So I do not believe we will ever run into the later case.

Reply:
So now, I do confirm the above is the original idea.....but double checking the latest version of the script, I see I have overlooked one bit and also the `agent_resource_control` and `upload_agent_config` step are triggered. While, I can see a reason why the later may be needed, I am not sure about the `agent_resource_control` though. We should discuss this. In the meantime I'll fix that with my next commit and will tie the resource control step to be dependent only on a separate `.dockerInit` file.

Reply:
As we a followup of our discussion on the topic, here is the issue created about refactoring this whole mechanism: https://github.com/dmwm/WMCore/issues/11627